Adds testing section (#71)
ember-engines/ember-engines.com: markdown/guide/testing.md

## Testing Engines

> > The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to maintain a separation of concerns between your app and engine, convert the engine into a separate addon.

Reviewer: ```suggestion
Just like other addons, engines must be tested differently depending upon whether they are "in-repo" or "standalone":
```

> > The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to create a separation of concerns between the app and your engine, extract the functionality into a normal addon.

Reviewer: ```suggestion
> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to maintain a separation of concerns between your app and engine, convert the engine into a separate addon.
```

> * To test an `in-repo-engine`, we must set up the tests from the host app

Reviewer: ```suggestion
* To test an in-repo engine, co-locate the engine's tests with the host app's tests, just like any other in-repo addon.
```

> * To test an `engine as an addon`, we must set up the tests from the dummy app

Reviewer: ```suggestion
* To test a standalone engine, use the dummy app, just like any other standalone addon project.
```

Ember Engines comes with a **set of test helpers that can be used in Unit/Integration tests**:

`engineResolverFor` gets the resolver class used by an Engine and creates an instance to be used with test modules:

```js
// tests/<integration or unit>/…
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';
```

### Unit/Integration Testing

To test candidates for unit/integration (e.g. components, services and controllers) declared inside an engine, you need to set a custom resolver with the engine's prefix using `engineResolverFor` helper.

> What does it look like to test a component from a host app or dummy app? Let's go over some examples in the next section. In the following tests, `admin-app` is an `in-repo-engine` or `engine as addon`, `hello-name` is a component, and `some-thing` is a service.

Reviewer: I think this is very confusing to call our engine `admin-app`, which makes it sound like an "app". How about `admin-console` or `admin-section` or even `admin-engine`?

Reviewer: Also, let's just use a single term `in-repo-engine` - saying `or 'engine as addon'` is super confusing.

Author: @dgeb how can we referrer here? `in-repo-engine` and `addon-engine`?

Reviewer: > `in-repo-engine` and `addon-engine`

Sure that's fine to distinguish between the two flavors of engines

My objection is that this phrase doesn't make sense:

> `admin-app` is an `in-repo-engine` or `engine as addon`

It has to be one or the other.

Author: fixed!

#### Unit Testing Basics

> Supposing that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.

Reviewer: ```suggestion
Suppose that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.
```

```js
// admin-app/services/some-thing-test.js
import Service from '@ember/service';
import { computed } from '@ember/object';

export default Service.extend({
  foo: 'bar',

  computedFoo: computed('foo', function() {
    return `computed ${this.foo}`;
  })
});
```

The unit test will be like this:

```js
// <app-name>/tests/unit/services/some-thing-test.js

import { module, test } from 'qunit';
import { setupTest } from 'ember-qunit';
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';

const modulePrefix = 'admin-app';
const resolver = engineResolverFor(modulePrefix);

module('Unit | Service | some thing', function(hooks) {
  setupTest(hooks, { resolver });

  test('should correctly concat foo', function(assert) {
    const someThing = this.owner.lookup('service:some-thing');
    someThing.set('foo', 'baz');

    assert.equal(someThing.get('computedFoo'), 'computed baz');
  });
});
```

#### Testing Components

Next, suppose that our engine has a component:

```hbs
{{!--  admin-app/addon/components/hello-name.hbs --}}
Hello, {{name}}!
```

> Here is what the integration test for that component might look:

Reviewer: ```suggestion
Here's an example integration test for that component:
```

```js
// <app-name>/tests/integration/components/hello-name-test.js

import { module, test } from 'qunit';
import { setupRenderingTest } from 'ember-qunit';
import { render } from '@ember/test-helpers';
import hbs from 'htmlbars-inline-precompile';
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';

const modulePrefix = 'admin-app';
const resolver = engineResolverFor(modulePrefix);

module('Integration | Component | hello-name', function(hooks) {
  setupRenderingTest(hooks, { resolver });

  test('it renders', async function(assert) {

    await render(hbs`{{#hello-name name="Tom"}}{{/hello-name}}`);

    assert.equal(this.element.textContent.trim(), 'Hello, Tom!');
  });
});
```

#### Acceptance Testing

Suppose that we are mouting `admin-app` on host-app router:

```js
// host-app/app/router.js
import EmberRouter from '@ember/routing/router';
import config from './config/environment';

const Router = EmberRouter.extend({
  location: config.locationType,
  rootURL: config.rootURL
});

Router.map(function() {
  this.mount('admin');
});

export default Router;
```

> Here is an acceptance test for routing::

Reviewer: ```suggestion
Here is an acceptance test for routing:
```

```js

import { module, test } from 'qunit';
import { setupApplicationTest } from 'ember-qunit';
import { visit, click, currentURL } from '@ember/test-helpers';

module('basic acceptance test', function(hooks) {
  setupApplicationTest(hooks);

  test('the user can visit /admin page', async function(assert) {
    await visit('/');
    await click('.admin-menu-item');

    assert.equal(currentURL(), '/admin');
  });
});
```

> ### Testing for Lazy Engines

Author: @rwjblue

> If you have a lazy engine, you'll need to edit your `tests/test-helper.js` like this:

Reviewer: ```suggestion
If you have a lazy engine, you'll need to ensure that your `tests/test-helper.js` is configured to preload your engine's assets:
```

```js
import Application from '../app';
import config from '../config/environment';
import { setApplication } from '@ember/test-helpers';
import { start } from 'ember-qunit';
import preloadAssets from 'ember-asset-loader/test-support/preload-assets';
import manifest from '<app-name>/config/asset-manifest';

setApplication(Application.create(config.APP));

preloadAssets(manifest).then(start); // This ensures all engine resources are loaded before the tests
```

> This should be enough to make unit, integration & acceptance tests work.

Reviewer: ```suggestion
This should be sufficient to make unit, integration, and acceptance tests work.
```

## Testing Engines

> > in-repo-engine & in-repo-addon doesn't have support to separate tests. If you want to create a separation of concerns between the app and your engine, to extract this into a normal addon.

Reviewer 2: ```suggestion
> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to create a separation of concerns between the app and your engine, extract the functionality into a normal addon.
```

> * for test an `in-repo-engine` we must setup the tests from the host app

Reviewer 2: ```suggestion
* To test an `in-repo-engine`, we must set up the tests from the host app
```

> * for test an `engine as addon` we must setup the tests from the dummy app

Reviewer 2: ```suggestion
* To test an `engine as an addon`, we must set up the tests from the dummy app
```

Ember Engines comes with a **set of test helpers that can be used in Unit/Integration tests**:

> `engineResolverFor` gets the resolver class used by an Engine and creates an instance to be used with test modules

Reviewer 2: ```suggestion
`engineResolverFor` gets the resolver class used by an Engine and creates an instance to be used with test modules:
```

```js
// tests/<integration or unit>/…
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';
```

### Unit/Integration Testing

To test candidates for unit/integration (e.g. components, services and controllers) declared inside an engine, you need to set a custom resolver with the engine's prefix using `engineResolverFor` helper.

> Assuming that we have an `in-repo-engine or engine as addon` called `admin-app` and it has a component `hello-name` and service `some-thing`. The following would be the setup to test such component from the `<app-name>` (host app or dummy app):

Reviewer 2: ```suggestion
What does it look like to test a component from a host app or dummy app? Let's go over some examples in the next section. In the following tests, `admin-app` is an `in-repo-engine` or `engine as addon`, `hello-name` is a component, and `some-thing` is a service.
```

#### Unit Testing Basics

> Supposing that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.

Reviewer 2: ```suggestion
In the first scenario, suppose that the engine has a service with a `computedFoo` computed property, based on a `foo` property:
```

```js
// admin-app/services/some-thing-test.js
import Service from '@ember/service';
import { computed } from '@ember/object';

export default Service.extend({
  foo: 'bar',

  computedFoo: computed('foo', function() {
    return `computed ${this.foo}`;
  })
});
```

> The unit test will be like this:

Reviewer 2: ```suggestion
Here is how we could unit test the service:
```

```js
// <app-name>/tests/unit/services/some-thing-test.js

import { module, test } from 'qunit';
import { setupTest } from 'ember-qunit';
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';

const modulePrefix = 'admin-app';
const resolver = engineResolverFor(modulePrefix);

module('Unit | Service | some thing', function(hooks) {
  setupTest(hooks, { resolver });

  test('should correctly concat foo', function(assert) {
    const someThing = this.owner.lookup('service:some-thing');
    someThing.set('foo', 'baz');

    assert.equal(someThing.get('computedFoo'), 'computed baz');
  });
});
```

#### Testing Components

> Supposing that we have in the engine this component

Reviewer 2: ```suggestion
Next, suppose that our engine has a component:
```

```hbs
{{!--  admin-app/addon/components/hello-name.hbs --}}
Hello, {{name}}!
```

> The integration test will be like this:

Reviewer 2: ```suggestion
Here is what the integration test for that component might look:
```

```js
// <app-name>/tests/integration/components/hello-name-test.js

import { module, test } from 'qunit';
import { setupRenderingTest } from 'ember-qunit';
import { render } from '@ember/test-helpers';
import hbs from 'htmlbars-inline-precompile';
import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';

const modulePrefix = 'admin-app';
const resolver = engineResolverFor(modulePrefix);

module('Integration | Component | hello-name', function(hooks) {
  setupRenderingTest(hooks, { resolver });

  test('it renders', async function(assert) {

    await render(hbs`{{#hello-name name="Tom"}}{{/hello-name}}`);

    assert.equal(this.element.textContent.trim(), 'Hello, Tom!');
  });
});
```

> **Note: you could create a helper and then use it like `Resolver from ../helpers/admin-app/resolver`**

Reviewer 2: I'm not sure what you mean by this note. Can you add some more detail here?

Author: fixed

#### Acceptance Testing

> Supposing that we are mouting `admin-app` on host-app router

Reviewer 2: ```suggestion
Suppose that we are mouting `admin-app` on host-app router:
```

```js
// host-app/app/router.js
import EmberRouter from '@ember/routing/router';
import config from './config/environment';

const Router = EmberRouter.extend({
  location: config.locationType,
  rootURL: config.rootURL
});

Router.map(function() {
  this.mount('admin');
});

export default Router;
```

> The acceptance test will be like this:

Reviewer 2: Here is an acceptance test for routing:

Author: fixed!
